iwyu.cc

```diff
@@ -4123,25 +4125,26 @@ class IwyuAstConsumer
     if (CanIgnoreCurrentASTNode())
       return true;
     // TypedefType::getDecl() returns the place where the typedef is defined.
-    if (CanForwardDeclareType(current_ast_node())) {
-      ReportDeclForwardDeclareUse(CurrentLoc(), type->getDecl());
-    } else {
-      ReportDeclUse(CurrentLoc(), type->getDecl());
-    }
+    ReportDeclUse(CurrentLoc(), type->getDecl());
```

Comment:
The old forward-declaration policy looks a little confused. We would only get here when a `typedef` type was used, right? And if we were in a forward-declare context, we would allow the `typedef` _declaration_ to be forward-declared (which essentially means duplicating it)?

On the other hand, I don't really understand why it's OK to omit `ReportTypeUse` if it's a forward-decl use? Ah, because if it's not a full use, this is just some implicit use of the type that we can ignore? Such as:
```
// blah.h
#include "some_typedef.h"
some_typedef_t* foo();

// main.cc
void bar() {
   (void)sizeof(foo());
}
```
?

Reply:
You're almost correct, except that in your case there should not be `TypedefType` in `bar()`'s AST at all, and `ReportTypeUse` is called by `sizeof` node handler. Forward-decl use in `VisitTypedefType` — it's about
`some_typedef_t* p = nullptr;`,
i.e. when typedef type is explicitly written.

Reply:
> And if we were in a forward-declare context, we would allow the typedef declaration to be forward-declared (which essentially means duplicating it)?

No, `TypedefDecl` fwd-decl uses are recategorized into full uses (see [`DeclCanBeForwardDeclared` logic](https://github.com/include-what-you-use/include-what-you-use/blob/0.19/iwyu_output.cc#L741)), so there is no difference between `ReportDeclUse` and `ReportDeclForwardDeclareUse` calls for `TypedefDecl`.

Reply:
Ah, OK, so that happens very late in the process, when doing the IWYU analysis. OK, fair enough. I think I see what's going on now.